runtime/vm/extendedMessageNPE.cpp

```diff
@@ -22,10 +22,227 @@
+ * to
+ * (java.lang.String[][][], java.lang.Void, byte, java.lang.String, int, char, double, float, long, java.lang.Object, short, boolean)
+ *
+ * Note:
+ * The return type is ignored.
+ * Assuming the method signature is well formed.
+ * The caller is responsible for freeing the returned string.
+ *
+ * @param[in] vmThread current J9VMThread
+ * @param[in] methodSig pointer to J9UTF8 containing the method signature
+ *
+ * @return a char pointer to the class name within fullQualifiedUTF,
+ *         NULL if not successful, but keep application exception instead of OOM.
+ */
+static char*
+convertMethodSignature(J9VMThread *vmThread, J9UTF8 *methodSig)
+{
+	UDATA i = 0;
+	UDATA j = 0;
+	UDATA arity = 0;
+	U_8 *string = J9UTF8_DATA(methodSig);
+	UDATA stringLength = J9UTF8_LENGTH(methodSig);
+	UDATA bufferSize = 0;
+	char *result = NULL;
+
+	PORT_ACCESS_FROM_VMC(vmThread);
+
+	/* first scan to calculate the buffer size required */
+	/* first character is '(' */
+	bufferSize += 1;
+	for (i = 1; ((')' != string[i]) && (i < stringLength)); i++) {
+		arity = 0;
+		while (('[' == string[i]) && (i < stringLength)) {
+			arity += 1;
+			i += 1;
+		}
+		switch (string[i]) {
+		case 'B': /* FALLTHROUGH */
+		case 'C': /* FALLTHROUGH */
+		case 'J':
+			/* byte, char, long */
+			bufferSize += 4;
+			break;
+		case 'D':
+			/* double */
+			bufferSize += 6;
+			break;
+		case 'F': /* FALLTHROUGH */
+		case 'S':
+			/* float, short */
+			bufferSize += 5;
+			break;
+		case 'I':
+			/* int */
+			bufferSize += 3;
+			break;
+		case 'L': {
+			i++;
+			UDATA objSize = 0;
+			while ((';' != string[i]) && (i < stringLength)) {
+				objSize++;
+				i++;
+			}
+			bufferSize += objSize;
+			break;
+		}
+		case 'Z':
+			/* boolean */
+			bufferSize += 7;
+			break;
+		default:
+			Trc_VM_ConvertMethodSignature_Malformed_Signature(vmThread, stringLength, string, i);
+			break;
+		}
+		bufferSize += (2 * arity);
+		if (')' != string[i + 1]) {
+			/* ", " */
+			bufferSize += 2;
+		}
+	}
+	/* ')' and the extra byte for '\0' */
+	bufferSize += 2;
+	Trc_VM_ConvertMethodSignature_Signature_BufferSize(vmThread, stringLength, string, bufferSize);
+	result = (char *)j9mem_allocate_memory(bufferSize, OMRMEM_CATEGORY_VM);
+	if (NULL != result) {
+		char *cursor = result;
+		UDATA availableSize = bufferSize;
+
+		memset(result, 0, bufferSize);
+		/* first character is '(' */
+		j9str_printf(PORTLIB, cursor, availableSize, "(");
+		cursor += 1;
+		availableSize -= 1;
+		for (i = 1; ((')' != string[i]) && (i < stringLength)); i++) {
+			arity = 0;
+			while (('[' == string[i]) && (i < stringLength)) {
+				arity += 1;
+				i += 1;
+			}
+			const char *elementType = NULL;
+			if ('L' == string[i]) {
+				i += 1;
+
+				UDATA objSize = 0;
+				while ((';' != string[i]) && (i < stringLength)) {
+					if ('/' == string[i]) {
+						*cursor = '.';
+					} else {
+						*cursor = string[i];
+					}
```

Comment:
This will be hot in cache, unconditionally copy it (common case), then overwrite with `.` in the uncommon case
```suggestion
				while ((';' != string[i]) && (i < stringLength)) {
					*cursor = string[i];
					if ('/' == string[i]) {
						*cursor = '.';
					} 
```

Reply:
Fixed.